File: lab/lab7/lab7/AVLTree.java
import java.util.Scanner;

public class AVLTree {
    static class Node {
        int data;
        int height;
        Node left;
        Node right;

        public Node(int value) {
            data = value;
            height = 1;
            left = null;
            right = null;
        }
    }

    private Node root;

    public AVLTree() {
        this.root = null;
    }

    private int getHeight(Node node) {
        if (node == null) {
            return 0;
        }
        return node.height;
    }

    private int getBalanceFactor(Node node) {
        if (node == null) {
            return 0;
        }
        return getHeight(node.left) - getHeight(node.right);
    }

    private void updateHeight(Node node) {
        if (node != null) {
            node.height = Math.max(getHeight(node.left), getHeight(node.right)) + 1;
        }
    }

    private Node rightRotate(Node y) {
        Node x = y.left;
        Node T2 = x.right;

        x.right = y;
        y.left = T2;

        updateHeight(y);
        updateHeight(x);

        return x;
    }

    private Node leftRotate(Node x) {
        Node y = x.right;
        Node T2 = y.left;

        y.left = x;
        x.right = T2;

        updateHeight(x);
        updateHeight(y);

        return y;
    }

    private Node insert(Node root, int value) {
        if (root == null) {
            return new Node(value);
        }

        if (value < root.data) {
            root.left = insert(root.left, value);
        } else if (value > root.data) {
            root.right = insert(root.right, value);
        } else {
            return root;
        }

        updateHeight(root);

        int balance = getBalanceFactor(root);

        if (balance > 1 && value < root.left.data) {
            return rightRotate(root);
        }

        if (balance < -1 && value > root.right.data) {
            return leftRotate(root);
        }

        if (balance > 1 && value > root.left.data) {
            root.left = leftRotate(root.left);
            return rightRotate(root);
        }

        if (balance < -1 && value < root.right.data) {
            root.right = rightRotate(root.right);
            return leftRotate(root);
        }

        return root;
    }

    public void insert(int value) {
        root = insert(root, value);
    }

    private void calculateAverageSearchLength(Node root, int depth, int[] nodesCount, int[] sumDepth) {
        if (root != null) {
            nodesCount[0]++;
            sumDepth[0] += depth;

            calculateAverageSearchLength(root.left, depth + 1, nodesCount, sumDepth);
            calculateAverageSearchLength(root.right, depth + 1, nodesCount, sumDepth);
        }
    }

    public void calculateAverageSearchLength() {
        int[] nodesCount = {0};
        int[] sumDepth = {0};
        calculateAverageSearchLength(root, 1, nodesCount, sumDepth);
        double asl = (double) sumDepth[0] / nodesCount[0];

        System.out.println("ASL：" + (asl - 1.0));
    }
    @Override
    public String toString() {
        return toStringHelper(root, "");
    }

    /* Recursive helper method for toString. */
    private String toStringHelper(Node node, String indent) {
        if (node == null) {
            return "";
        } else {
            String toReturn = "";
            Node rightChild = node.right;
            toReturn += toStringHelper(rightChild, "        " + indent);
            if (rightChild != null) {
                toReturn += indent + "    /";
            }
            //toReturn += "\n" + indent + node.data + " (Balance Factor: " + getBalanceFactor(node) + ")\n";
            toReturn += "\n" + indent + node.data + "\n";
            Node leftChild = node.left;
            if (leftChild != null) {
                toReturn += indent + "    \\";
            }
            toReturn += toStringHelper(leftChild, "        " + indent);
            return toReturn;
        }
    }

    public static void main(String[] args) {
        AVLTree avlTree = new AVLTree();

        // Input a sequence of integers and generate an AVL tree
        Scanner scanner = new Scanner(System.in);
        System.out.println("Enter a sequence of integers (end with Enter):");

        while (scanner.hasNextLine()) {
            String line = scanner.nextLine().trim();
            if (line.isEmpty()) {
                // Stop input if an empty line is entered
                break;
            }
            try {
                int num = Integer.parseInt(line);
                avlTree.insert(num);


                //Display each step
                System.out.println("AVL Tree representation:");
                System.out.println(avlTree);

            } catch (NumberFormatException e) {
                System.out.println("Invalid input. Please enter integers only.");
            }
        }

        // Display the AVL Tree
        System.out.println("AVL Tree representation:");
        System.out.println(avlTree);

        avlTree.calculateAverageSearchLength();
    }
}

File: hw/hw3/AVLtree.cpp
#include <iostream>
#include <cmath>

using namespace std;

// 二叉排序树的节点
struct TreeNode {
    int data;
    int height; // 节点的高度
    TreeNode* left;
    TreeNode* right;

    TreeNode(int value) : data(value), height(1), left(nullptr), right(nullptr) {}
};

// 获取节点的高度
int getHeight(TreeNode* node) {
    if (node == nullptr) {
        return 0;
    }
    return node->height;
}

// 获取节点的平衡因子
int getBalanceFactor(TreeNode* node) {
    if (node == nullptr) {
        return 0;
    }
    return getHeight(node->left) - getHeight(node->right);
}

// 更新节点的高度
void updateHeight(TreeNode* node) {
    if (node != nullptr) {
        node->height = max(getHeight(node->left), getHeight(node->right)) + 1;
    }
}

// 右旋操作
TreeNode* rightRotate(TreeNode* y) {
    TreeNode* x = y->left;
    TreeNode* T2 = x->right;

    // 执行右旋
    x->right = y;
    y->left = T2;

    // 更新节点高度
    updateHeight(y);
    updateHeight(x);

    return x;
}

// 左旋操作
TreeNode* leftRotate(TreeNode* x) {
    TreeNode* y = x->right;
    TreeNode* T2 = y->left;

    // 执行左旋
    y->left = x;
    x->right = T2;

    // 更新节点高度
    updateHeight(x);
    updateHeight(y);

    return y;
}

// 插入节点，并保持树平衡
TreeNode* insert(TreeNode* root, int value) {
    // 执行标准BST插入
    if (root == nullptr) {
        return new TreeNode(value);
    }

    if (value < root->data) {
        root->left = insert(root->left, value);
    } else if (value > root->data) {
        root->right = insert(root->right, value);
    } else {
        // 重复元素不插入
        return root;
    }

    // 更新节点高度
    updateHeight(root);

    // 获取平衡因子
    int balance = getBalanceFactor(root);

    // 左子树不平衡，右旋操作
    if (balance > 1 && value < root->left->data) {
        return rightRotate(root);
    }

    // 右子树不平衡，左旋操作
    if (balance < -1 && value > root->right->data) {
        return leftRotate(root);
    }

    // 左右不平衡，先左旋后右旋操作
    if (balance > 1 && value > root->left->data) {
        root->left = leftRotate(root->left);
        return rightRotate(root);
    }

    // 右左不平衡，先右旋后左旋操作
    if (balance < -1 && value < root->right->data) {
        root->right = rightRotate(root->right);
        return leftRotate(root);
    }

    return root;
}

// 计算平均查找长度
void calculateAverageSearchLength(TreeNode* root, int depth, int& nodesCount, int& sumDepth) {
    if (root != nullptr) {
        nodesCount++;
        sumDepth += depth;

        calculateAverageSearchLength(root->left, depth + 1, nodesCount, sumDepth);
        calculateAverageSearchLength(root->right, depth + 1, nodesCount, sumDepth);
    }
}

int main() {
    // 数列L
    int arr[] = {1, 2, 3, 4, 5, 6, 7, 8, 9};
    int n = sizeof(arr) / sizeof(arr[0]);

    TreeNode* root = nullptr;

    // 插入元素构建AVL树
    for (int i = 0; i < n; i++) {
        root = insert(root, arr[i]);    
    }

    // 计算平均查找长度
    int nodesCount = 0;
    int sumDepth = 0;
    calculateAverageSearchLength(root, 1, nodesCount, sumDepth);

    // 输出结果
    cout << "平均查找长度：" << static_cast<double>(sumDepth) / nodesCount << endl;

    return 0;
}
